lms/djangoapps/instructor/permissions.py

```diff
@@ -44,7 +44,13 @@
 # only global staff or those with the data_researcher role can access the data download tab
 # HasAccessRule('staff') also includes course staff
 perms[CAN_RESEARCH] = is_staff | HasRolesRule('data_researcher')
-perms[CAN_ENROLL] = HasAccessRule('staff')
+# eshe_instructor implicitly gets staff access, but shouldn't be able to enroll
+perms[CAN_ENROLL] = (
+    # can enroll if a user is an eshe_instructor and has an explicit staff role
+    (HasRolesRule('eshe_instructor') & HasAccessRule('staff', strict=True)) |
+    # can enroll if a user is just staff
+    (~HasRolesRule('eshe_instructor') & HasAccessRule('staff'))
+)
```

Comment:
@0x29a Could you explain this more in depth for me? I don't understand why we can't set it to `HasAccessRule('staff')`, and do we need to check with no inheritance for the `HasAccessRule('staff', strict=True)`, since `staff` doesn't inherit from the `eshe_instructor` role, it's the other way around.

Reply:
> and do we need to check with no inheritance for the HasAccessRule('staff', strict=True), since staff doesn't inherit from the eshe_instructor role, it's the other way around.

Never mind, I got confused. I realize that we need to check strictly because `HasAccessRule('staff')` can return `true` if someone has `eshe_instructor` role, since we are inheriting. And the rest is so we don't break the inheritance for the other roles that inherit from the `staff`.

Reply:
Yep, that's right @Cup0fCoffee. Actually, I'm not sure that `explicit` / `implicit` wording in comments explaining this are the best choice. I'll think how I can explain this workaround more clearly.